### PR TITLE
Deprecate output and positional args in `filters.gaussian`

### PR DESCRIPTION
**Pull request by lagru**
## Description

Addresses an item in our [skimage2 transition list](https://github.com/scikit-image/scikit-image/wiki/API-changes-for-skimage2). We can't simply use `deprecate_kwarg` or `remove_arg`, as `output` can also be passed as a positional argument. Instead, deprecate passing more than 1 positional argument and at the same time, refactor `output` to `out`.

I'm not really sure whether `remove_arg` is a user friendly way to deprecate a positional parameter. Applying the function prevents positional arguments after the deprecated one as well, even though this is not deprecated. The warning will users also not tell them about it. I find it clearer to just deprecate these positional arguments in favor of keyword-only ones. I also think it's the better practice.

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
Deprecate parameter `output` in `skimage.filters.gaussian`; use `out` instead.
```

### CODE REVIEW

File: skimage/_shared/filters.py [hunk 1]
```diff
@@ -4,14 +4,47 @@
...
 
 from .._shared.utils import _supported_float_type, convert_to_float
 
 
+def _deprecate_gaussian_positional_args(func):
+    """Help deprecate `output` in `gaussian` in favor of `out`.
+
+    We can't simply use deprecate_kwarg or remove_arg, as `output` can also be
+    passed as a positional argument. Instead, deprecate passing more than 1
+    positional argument and at the same time, refactor `output` to `out`.
```
Comment (mkcor) [inline skimage/_shared/filters.py:22]:
```suggestion
    positional argument and, at the same time, refactor `output` to `out`.
```

File: skimage/_shared/filters.py [hunk 2]
```diff
@@ -4,14 +4,47 @@
...
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        if len(args) > 1:
+            warnings.warn(
+                "Passing more than 1 positional argument to `gaussian` is deprecated "
+                "and will stop working in version 0.25 or later. Please use keyword "
```
Comment (mkcor) [inline skimage/_shared/filters.py:30]:
```suggestion
                "and will stop working in version 0.25. Please use keyword "
```
Alternatively, "... will not work in versions 0.25 and later."

Unless I didn't get the meaning right, and you want to leave it open: We will decide later when it will stop working (either in version 0.25 or in a later version). But I'd rather communicate a definite version for when it stops working.

File: skimage/_shared/filters.py [hunk 3]
```diff
@@ -4,14 +4,47 @@
...
+                "arguments instead.",
+                FutureWarning,
+                stacklevel=2,
+            )
+        elif "output" in kwargs:
+            warnings.warn(
+                "`output` is a deprecated argument name for `gaussian`. It will be "
+                "removed in version 0.25 or later. Please use `out` instead.",
```
Comment (mkcor) [inline skimage/_shared/filters.py:38]:
```suggestion
                "removed in version 0.25. Please use `out` instead.",
```
(likewise)

File: skimage/_shared/filters.py [hunk 4]
```diff
@@ -22,9 +55,16 @@ def gaussian(
...
     *,
     channel_axis=None,
+    out=None,
 ):
     """Multi-dimensional Gaussian filter.
 
+    .. warning::
+
+        Passing more than 1 positional argument to `gaussian` is deprecated
+        and will stop working in version 0.25 or later. Please use keyword
```
Comment (mkcor) [inline skimage/_shared/filters.py:65]:
```suggestion
        and will stop working in version 0.25. Please use keyword
```

File: skimage/_shared/filters.py [hunk 5]
```diff
@@ -34,9 +74,8 @@ def gaussian(
         deviations of the Gaussian filter are given for each axis as a
         sequence, or as a single number, in which case it is equal for
         all axes.
-    output : ndarray, optional
-        The ``output`` parameter passes an array in which to store the
-        filter output.
+    output : DEPRECATED
+        Deprecated in favor of `out`
```
Comment (mkcor) [inline skimage/_shared/filters.py:78]:
```suggestion
        Deprecated in favor of `out`.
```

File: skimage/_shared/filters.py [hunk 6]
```diff
@@ -61,16 +100,12 @@ def gaussian(
 
         .. versionadded:: 0.19
            ``channel_axis`` was added in 0.19.
+    out : ndarray, optional
+        The ``out`` parameter passes an array in which to store the
```
Comment (mkcor) [inline skimage/_shared/filters.py:104]:
```suggestion
        The `out` parameter passes an array in which to store the
```
https://github.com/scikit-image/scikit-image/pull/7154#discussion_r1374763766 :sweat_smile: 

Also, I would reword more lightly: "The `out` parameter is for storing the filtered array (output)."

File: skimage/_shared/filters.py [hunk 7]
```diff
@@ -134,8 +171,8 @@ def gaussian(
     image = convert_to_float(image, preserve_range)
     float_dtype = _supported_float_type(image.dtype)
     image = image.astype(float_dtype, copy=False)
-    if (output is not None) and (not np.issubdtype(output.dtype, np.floating)):
-        raise ValueError("Provided output data type is not float")
+    if (out is not None) and (not np.issubdtype(out.dtype, np.floating)):
+        raise ValueError("Provided out data type is not float")
```
Comment (mkcor) [inline skimage/_shared/filters.py:175]:
```suggestion
        raise ValueError("Provided `out` data type is not float")
```
Works better for the user reading their screen.

File: skimage/_shared/filters.py [hunk 8]
```diff
@@ -83,8 +118,8 @@ def gaussian(
 
     Integer arrays are converted to float.
 
-    The ``output`` should be floating point data type since gaussian converts
-    to float provided ``image``. If ``output`` is not provided, another array
+    `out` should be a floating point data type since gaussian converts
```
Comment (mkcor) [inline skimage/_shared/filters.py:121]:
```suggestion
    `out` should be of floating-point data type since `gaussian` converts
```
Here, 'data type' means dtype.

File: skimage/_shared/filters.py [hunk 9]
```diff
@@ -83,8 +118,8 @@ def gaussian(
...
+    to float provided `image`. If `out` is not provided, another array
```
Comment (mkcor) [inline skimage/_shared/filters.py:122]:
```suggestion
    input `image` to float. If `out` is not provided, another array
```

File: skimage/filters/tests/test_gaussian.py [hunk 10]
```diff
@@ -164,3 +164,34 @@ def test_dog_invalid_sigma2():
         difference_of_gaussians(image, 3, 2)
     with pytest.raises(ValueError):
         difference_of_gaussians(image, (1, 5), (2, 4))
+
+
+@pytest.mark.parametrize("positional_count", range(1, 5))
+def test_deprecated_positional_args(positional_count):
+    """More than 1 positionl argument are deprecated."""
```
Comment (mkcor) [inline skimage/filters/tests/test_gaussian.py:171]: I would remove this 'docstring' since you didn't use one in the next test -- maybe because you find the function name to be descriptive enough? And the string to match (line 177) is more informative than this docstring.

**Review by mkcor — COMMENTED** [review]
Nice work, @lagru! Thanks for tackling one item in the list and doing the heavy lifting...

**Comment by lagru** [comment]
Thanks for the review @mkcor. I am currently working on merging and updating our `remove_arg` and `deprecate_kwarg` decorators. IMO we can combine them into a simpler `deprecate_parameter` that allows replacing positional parameters as well. That's currently more awkward then necessary. And we will need to do so for skimage2 quite a bit.

After that I'll address your feedback and update this PR accordingly.

**Comment by lagru** [comment]
@mkcor, sorry for the "Marked as resolved / unresolved" spam, I got confused. :see_no_evil: 

This should be ready to review / merge.

**Comment by lagru** [comment]
@scikit-image/core, this would be good to prioritize as a step towards skimage2. :pray:

File: skimage/_shared/filters.py [hunk 11]
```diff
@@ -61,16 +67,11 @@ def gaussian(
 
         .. versionadded:: 0.19
            ``channel_axis`` was added in 0.19.
+    out : ndarray, optional
+        If given, the filtered array will be stored in this array.
```
Comment (mkcor) [inline skimage/_shared/filters.py:71]:
```suggestion
        If given, the filtered image will be stored in this array.
```

File: skimage/_shared/filters.py [hunk 12]
```diff
@@ -61,16 +67,11 @@ def gaussian(
 
         .. versionadded:: 0.19
            ``channel_axis`` was added in 0.19.
```
Comment (mkcor) [inline skimage/_shared/filters.py:69]:
```suggestion
           `channel_axis` was added in 0.19.
```
En passant, consistently with line 74.

File: skimage/_shared/filters.py [hunk 13]
```diff
@@ -83,8 +84,8 @@ def gaussian(
 
     Integer arrays are converted to float.
 
-    The ``output`` should be floating point data type since gaussian converts
-    to float provided ``image``. If ``output`` is not provided, another array
+    `out` should be of floating-point data type since `gaussian` converts the
+    provided `image` to float. If `out` is not provided, another array
```
Comment (mkcor) [inline skimage/_shared/filters.py:88]:
```suggestion
    input `image` to float. If `out` is not provided, another array
```

File: skimage/_shared/filters.py [hunk 14]
```diff
@@ -134,8 +134,8 @@ def gaussian(
     image = convert_to_float(image, preserve_range)
     float_dtype = _supported_float_type(image.dtype)
     image = image.astype(float_dtype, copy=False)
-    if (output is not None) and (not np.issubdtype(output.dtype, np.floating)):
-        raise ValueError("Provided output data type is not float")
+    if (out is not None) and (not np.issubdtype(out.dtype, np.floating)):
+        raise ValueError(f"dtype of `out` must be a floating type, was {out.dtype!r}")
```
Comment (mkcor) [inline skimage/_shared/filters.py:138]: Sorry, iterating here:
```suggestion
        raise ValueError(f"dtype of `out` must be float; provided {out.dtype!r}.")
```
Either "`out` should be of floating-point dtype" or "dtype of `out` should be float" (right?).

**Review by mkcor — APPROVED** [review]
Great work, @lagru! Only left minor suggestions.

File: skimage/_shared/filters.py (hunk 14 above)
Comment (lagru) [reply skimage/_shared/filters.py]: Addressed in https://github.com/scikit-image/scikit-image/pull/7225/commits/f01f9085e261e1989fe453a23d5526a3a495260c. "I felt like "got" conveyed the same meaning as "provided" but is a bit less advanced vocabulary."

**Comment by soupault** [comment]
Thank you, Lars!

>deprecate passing more than 1 positional argument

If it is still planned for the current branch, I think it would be acceptable to have both deprecations running at the same time. The implementation could then count the number of arguments passed positionally (e.g. as in https://stackoverflow.com/questions/67639234/count-how-many-arguments-passed-as-positional).

No comments to the PR otherwise, and I am happy to proceed with the current version.

**Comment by lagru** [comment]
Good call @soupault. I forgot to update the original intention in the PR description. I originally wanted to turn `sigma` into a keyword-only parameter as well but then decided to focus on the actual goal of the PR (I think).

I'd be happy if we turned that parameter into a keyword-only one but think it would be reasonable to do it in another PR perhaps?

**Comment by soupault** [comment]
Sure. Thanks for working on this!
